### PR TITLE
fix: Respecting max/min opacities, and adding tests.

### PR DESCRIPTION
**Pull request by rusackas**
### SUMMARY
There was an error where calculated opacities could be over 1, or below 0. This code will only allow the opacity to go to the configured maximum or minimum. 

Hat tip to @michael-s-molina for the detective work here :)

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

image

### TESTING INSTRUCTIONS

Unit tests included!

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/packages/superset-ui-chart-controls/test/utils/getColorFormatters.test.ts [hunk 1]
```diff
@@ -50,6 +50,8 @@ describe('getOpacity', () => {
     expect(getOpacity(100, 100, 50)).toEqual(0.05);
     expect(getOpacity(100, 100, 100, 0, 0.8)).toEqual(0.8);
     expect(getOpacity(100, 100, 50, 0, 1)).toEqual(0);
+    expect(getOpacity(999, 100, 50, 0, 1)).toEqual(1);
+    expect(getOpacity(-999, 100, 50, 0, 1)).toEqual(0.05);
```
Comment (michael-s-molina) [inline superset-frontend/packages/superset-ui-chart-controls/test/utils/getColorFormatters.test.ts:54]: `Math.abs` inside the function will make sure `result` is never negative. We can replace this test with one that passes a high `minOpacity` like `0.9` and parameters that make `result` be less than this value to test `result < minOpacity`
```suggestion
```

**Review by michael-s-molina — APPROVED** [review]
LGTM

File: superset-frontend/packages/superset-ui-chart-controls/test/utils/getColorFormatters.test.ts [hunk 2]
```diff
@@ -50,6 +50,8 @@ describe('getOpacity', () => {
     expect(getOpacity(100, 100, 50)).toEqual(0.05);
     expect(getOpacity(100, 100, 100, 0, 0.8)).toEqual(0.8);
     expect(getOpacity(100, 100, 50, 0, 1)).toEqual(0);
+    expect(getOpacity(999, 100, 50, 0, 1)).toEqual(1);
+    expect(getOpacity(100, 100, 50, 0.99, 1)).toEqual(0.99);
```
Comment (rusackas) [inline superset-frontend/packages/superset-ui-chart-controls/test/utils/getColorFormatters.test.ts:54]:
```suggestion
    expect(getOpacity(100, 100, 50, 0.99, 1)).toEqual(0.99);
    expect(getOpacity(.01, 100, 50, 0.05, 1)).toEqual(0.05);
```

**Comment by sadpandajoe** [comment]
🏷️ preset:2022.25

File: superset-frontend/packages/superset-ui-chart-controls/src/utils/getColorFormatters.ts [hunk 3]
```diff
@@ -37,16 +37,24 @@ export const getOpacity = (
...
+        (value - cutoffPoint),
+    ) + minOpacity,
+    2,
+  );
+
+  if (result < minOpacity) {
+    return minOpacity;
+  }
+
+  return result > maxOpacity ? maxOpacity : result;
```
Comment (stephenLYZ) [inline superset-frontend/packages/superset-ui-chart-controls/src/utils/getColorFormatters.ts]: We could probably use one line of code
```javascript
// Make sure that the opacity is within the bounds
return Math.min(Math.max(result, minOpacity), maxOpacity)
```

**Pull request merged**